v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -239,32 +249,88 @@ async def set_disconnected():
                 # Paho handler invoke until we know the connection state has been set.
                 f.result()
 
-            if do_stop_network_loop:
-
-                async def stop_network_loop():
-                    logger.debug("Stopping Paho network loop due to unexpected disconnect")
-                    client.loop_stop()
+                # Cancel pending subscribes and unsubscribes only.
+                # Publishes can survive a disconnect.
+                async def cancel_pending():
+                    if len(this._pending_subs) != 0 or len(this._pending_unsubs) != 0:
+                        async with this._mid_tracker_lock:
+                            logger.debug("Cancelling pending subscribes")
+                            mids = this._pending_subs.keys()
+                            for mid in mids:
+                                this._pending_subs[mid].cancel()
+                            this._pending_subs.clear()
+                            logger.debug("Cancelling pending unsubscribes")
+                            mids = this._pending_unsubs.keys()
+                            for mid in mids:
+                                this._pending_unsubs[mid].cancel()
+                            this._pending_unsubs.clear()
+
+                # NOTE: This coroutine might not be able to finish right away due to the
+                # mid_tracker_lock. Don't wait on it's completion or it may lock up.
+                asyncio.run_coroutine_threadsafe(cancel_pending(), this._loop)
+
+        def on_subscribe(client, userdata, mid, granted_qos):
+            this = self_weakref()
+            logger.debug("SUBACK received for mid {}".format(mid))
 
-                # NOTE: This coroutine can't finish execution until after this handler exits
-                # since this handler itself is running on the network loop.
-                # This means this MUST be the last task that gets scheduled in this handler,
-                # since .loop_stop() will block until this handler finishes running.
-                asyncio.run_coroutine_threadsafe(stop_network_loop(), this._loop)
+            async def complete_sub():
+                async with this._mid_tracker_lock:
+                    try:
+                        f = this._pending_subs[mid]
+                        f.set_result(True)
+                    except KeyError:
+                        logger.warning("Unexpected SUBACK received for mid {}".format(mid))
 
-        def on_subscribe(client, userdata, mid):
-            pass
+            asyncio.run_coroutine_threadsafe(complete_sub(), this._loop)
```

Comment:
>asyncio.run_coroutine_threadsafe(complete_sub(), this._loop) [](http://example.com/codeflow?start=12&length=60)

It's super important that you don't call result() on the Future that is returned.

Reply:
Yeah, I should have documented that, good call.